File: src/Tests/Behavioral/DerefPointerToField/DerefPointerToField.go
package main

import "fmt"

type holder struct {
	xs  *[]int // pointer to a slice
	cnt *int   // pointer to a value
}

// sumRange ranges over a deref'd pointer-to-slice FIELD of a pointer parameter (`*h.xs`). The deref
// must apply to the field, not be dropped (the param-deref shortcut wrongly matched `*h.field`).
func sumRange(h *holder) int {
	s := 0
	for _, x := range *h.xs {
		s += x
	}
	return s
}

// readVal dereferences a pointer FIELD of a pointer parameter (`*h.cnt`) for a value read.
func readVal(h *holder) int {
	return *h.cnt
}

func main() {
	xs := []int{10, 20, 30}
	c := 7
	h := &holder{xs: &xs, cnt: &c}
	fmt.Println(sumRange(h)) // 60
	fmt.Println(readVal(h))  // 7
}

File: src/Tests/Behavioral/DerefPointerToField/DerefPointerToField.cs
namespace go;

using fmt = fmt_package;

partial class main_package {

[GoType] partial struct holder {
    internal ж<slice<nint>> xs;
    internal ж<nint> cnt;
}

internal static nint sumRange(ж<holder> Ꮡh) {
    ref var h = ref Ꮡh.Value;

    nint s = 0;
    foreach (var (_, x) in h.xs.ValueSlot) {
        s += x;
    }
    return s;
}

internal static nint readVal(ж<holder> Ꮡh) {
    ref var h = ref Ꮡh.Value;

    return h.cnt.Value;
}

internal static void Main() {
    var xs = new nint[]{10, 20, 30}.slice();
    ref var c = ref heap<nint>(out var Ꮡc);
    c = 7;
    var h = Ꮡ(new holder(xs: Ꮡ(xs), cnt: Ꮡc));
    fmt.Println(sumRange(h));
    fmt.Println(readVal(h));
}

} // end main_package
